examples/fabric-admin/commands/pairing/DeviceSynchronization.cpp

```diff
@@ -0,0 +1,176 @@
+        break;
+    case Clusters::BasicInformation::Attributes::HardwareVersionString::Id:
+        mCurrentDeviceData.has_hardware_version_string = SuccessOrLog(
+            data->GetString(mCurrentDeviceData.hardware_version_string, sizeof(mCurrentDeviceData.hardware_version_string)),
+            "HardwareVersionString");
+        break;
+    case Clusters::BasicInformation::Attributes::SoftwareVersion::Id:
+        mCurrentDeviceData.has_software_version = SuccessOrLog(data->Get(mCurrentDeviceData.software_version), "HardwareVersion");
+        break;
+    case Clusters::BasicInformation::Attributes::SoftwareVersionString::Id:
+        mCurrentDeviceData.has_software_version_string = SuccessOrLog(
+            data->GetString(mCurrentDeviceData.software_version_string, sizeof(mCurrentDeviceData.software_version_string)),
+            "SoftwareVersionString");
+        break;
+    default:
+        break;
+    }
+}
+
+void DeviceSynchronizer::OnReportEnd()
+{
+    // Report end is at the end of all attributes (success)
+#if defined(PW_RPC_ENABLED)
+    AddSynchronizedDevice(mCurrentDeviceData);
+#else
+    ChipLogError(NotSpecified, "Cannot synchronize device with fabric bridge: RPC not enabled");
+#endif
+}
+
+void DeviceSynchronizer::OnDone(chip::app::ReadClient * apReadClient)
+{
+    // Nothing to do: error reported on OnError or report ended called.
+}
+
+void DeviceSynchronizer::OnError(CHIP_ERROR error)
+{
+    ChipLogProgress(NotSpecified, "Error fetching device data: %" CHIP_ERROR_FORMAT, error.Format());
+}
+
+void DeviceSynchronizer::OnDeviceConnected(chip::Messaging::ExchangeManager & exchangeMgr,
+                                           const chip::SessionHandle & sessionHandle)
+{
+    mClient = std::make_unique<ReadClient>(app::InteractionModelEngine::GetInstance(), &exchangeMgr /* echangeMgr */,
+                                           *this /* callback */, ReadClient::InteractionType::Read);
+    VerifyOrDie(mClient);
+
+    AttributePathParams readPaths[1];
+    readPaths[0] = AttributePathParams(kRootEndpointId, Clusters::BasicInformation::Id);
+
+    ReadPrepareParams readParams(sessionHandle);
+
+    readParams.mpAttributePathParamsList    = readPaths;
+    readParams.mAttributePathParamsListSize = 1;
+
+    CHIP_ERROR err = mClient->SendRequest(readParams);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to issue read for BasicInformation data");
+    }
+}
+
+void DeviceSynchronizer::OnDeviceConnectionFailure(const chip::ScopedNodeId & peerId, CHIP_ERROR error)
+{
+    ChipLogError(NotSpecified, "Device Sync failed to connect to " ChipLogFormatX64, ChipLogValueX64(peerId.GetNodeId()));
+}
+
+void DeviceSynchronizer::StartDeviceSynchronization(chip::Controller::DeviceController & controller, chip::NodeId nodeId,
+                                                    bool deviceIsIcd)
+{
+    mCurrentDeviceData            = chip_rpc_SynchronizedDevice_init_default;
```

Comment:
Does it make sense to have a bool mOngoingDeviceSync where we do VerifyOrDie(!mOngoingDeviceSync). With a comment saying this currently assumes only one ongoing device sync at a time

mOngoingDeviceSync is set to true here, and set to false in `OnReportEnd` or `OnDone`

Reply:
Added, set to false on ReportEnd/OnError since those seem to be the final calls.